cpp/src/strings/copy/repeat_join.cu

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/get_value.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/strings/copy.hpp>
+#include <cudf/strings/detail/utilities.cuh>
+#include <cudf/strings/detail/utilities.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+#include <cudf/utilities/error.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+
+namespace cudf {
+namespace strings {
+namespace detail {
+
+string_scalar repeat_join(string_scalar const& input,
+                          size_type repeat_times,
+                          rmm::cuda_stream_view stream,
+                          rmm::mr::device_memory_resource* mr)
+{
+  if (!input.is_valid()) { return string_scalar("", false, stream, mr); }
+  if (input.size() == 0 || repeat_times <= 0) { return string_scalar("", true, stream, mr); }
+  if (repeat_times == 1) { return string_scalar(input, stream, mr); }
+
+  CUDF_EXPECTS(input.size() <= std::numeric_limits<size_type>::max() / repeat_times,
+               "The output string has size that exceeds the maximum allowed size.");
+
+  auto buff =
+    rmm::device_buffer(repeat_times * input.size(), stream, rmm::mr::get_current_device_resource());
+  for (size_type i = 0, str_size = input.size(); i < repeat_times; ++i) {
+    auto const offset = i * str_size;
+    thrust::copy(rmm::exec_policy(stream),
+                 input.data(),
+                 input.data() + str_size,
+                 static_cast<char*>(buff.data()) + offset);
+  }
```

Comment:
This seems fairly inefficient. How big is `repeat_times` expected to be? 
Since the input is a fixed length, seems we could better parallelize this.
Also, it seems unnecessary to copy the device buffer.

Reply:
Yes, it seems to be inefficient. Do you have any suggestions for this? I don't know if we can do this by any `thrust::fill` API.

Reply:
You can't use fill because you have an array of bytes (device_buffer) and the bytes are not all the same. 

To achieve good parallelism, have each thread copy the appropriate character from input.data(). That is, output_chars[i] = input_chars[i % str_size]. You can do this in a thrust::transform with a counting iterator.